src/Estimators/InputSection.cpp

```diff
@@ -0,0 +1,176 @@
+//////////////////////////////////////////////////////////////////////////////////////
+// This file is distributed under the University of Illinois/NCSA Open Source License.
+// See LICENSE file in top directory for details.
+//
+// Copyright (c) 2021 QMCPACK developers.
+//
+// File developed by: Jaron T. Krogel, krogeljt@ornl.gov, Oak Ridge National Laboratory
+//
+//////////////////////////////////////////////////////////////////////////////////////
+
+#include "InputSection.h"
+#include "Message/UniformCommunicateError.h"
+
+namespace qmcplusplus
+{
+
+void InputSection::readXML(xmlNodePtr cur)
+{
+  // read attributes
+  xmlAttrPtr att = cur->properties;
+  while (att != NULL)
+  {
+    std::string name = (const char*)(att->name);
+    for (auto& c : name)
+      c = tolower(c);
+    if (!is_attribute(name))
+    {
+      std::stringstream error;
+      error << "InputSection::readXML name " << name << " is not an attribute of " << section_name << "\n";
+      throw UniformCommunicateError(error.str());
+    }
+    std::istringstream stream((const char*)(att->children->content));
+    set_from_stream(name, stream);
+    att = att->next;
+  }
+
+  // read parameters
+  xmlNodePtr element = cur->xmlChildrenNode;
+  while (element != NULL)
+  {
+    std::string ename((const char*)element->name);
+    for (auto& c : ename)
+      c = tolower(c);
+    if (ename == "parameter")
+    {
+      XMLAttrString name(element, "name");
+      for (auto& c : name)
+        c = tolower(c);
+      if (!is_parameter(name))
+      {
+        std::stringstream error;
+        error << "InputSection::readXML name " << name << " is not a parameter of " << section_name << "\n";
+        throw UniformCommunicateError(error.str());
+      }
+      std::istringstream stream(XMLNodeString{element});
+      set_from_stream(name, stream);
+    }
+    element = element->next;
+  }
+
+  // assign default values for optional variables
+  set_defaults();
+
+  // check input validity
+  check_valid();
+
+  //report();
+};
+
+
+void InputSection::init(const std::unordered_map<std::string, std::any>& init_values)
+{
+  for (auto& [name, value] : init_values)
+    if (is_string(name))
+      values[name] = std::any_cast<std::string>(value);
+    else if (is_bool(name))
+      values[name] = std::any_cast<bool>(value);
+    else if (is_integer(name))
+      values[name] = std::any_cast<int>(value);
+    else if (is_real(name))
+      values[name] = std::any_cast<Real>(value);
+    else
+    {
+      std::stringstream error;
+      error << "InputSection::init name " << name << " in " << section_name << " does not have a type\n";
+      throw UniformCommunicateError(error.str());
+    }
+
+  // assign default values for optional variables
+  set_defaults();
+
+  // check input validity
+  check_valid();
+
+  //report();
+};
+
+
+void InputSection::report() const
```

Comment:
Needs testing someone may mess with the is_xxx in the future this has a bunch that can break.

If you imagine yourself or anyone parsing this output you should test and therefor specify which parts are "guaranteed" to be conserved.

Reply:
This is intended primarily as a developer/debugging function and it called nowhere in the main code.  I will mark it as such with additional comments and move it to private so as not to confuse it with the user API.